cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,530 @@
+            auto hist_scan_deps = radix_hist_scan(queue_,
+                                                  part_hist_,
+                                                  part_prefix_hist_,
+                                                  local_size_,
+                                                  local_hist_count_,
+                                                  scan_deps);
+            res_deps = radix_reorder(queue_,
+                                     val_in,
+                                     ind_in,
+                                     part_prefix_hist_,
+                                     val_buff_,
+                                     ind_buff_,
+                                     elem_count_,
+                                     bit_offset,
+                                     local_size_,
+                                     local_hist_count_,
+                                     hist_scan_deps);
+        }
+        else {
+            auto scan_deps = radix_scan(queue_,
+                                        val_buff_,
+                                        part_hist_,
+                                        elem_count_,
+                                        bit_offset,
+                                        local_size_,
+                                        local_hist_count_,
+                                        res_deps);
+            auto hist_scan_deps = radix_hist_scan(queue_,
+                                                  part_hist_,
+                                                  part_prefix_hist_,
+                                                  local_size_,
+                                                  local_hist_count_,
+                                                  scan_deps);
+            res_deps = radix_reorder(queue_,
+                                     val_buff_,
+                                     ind_buff_,
+                                     part_prefix_hist_,
+                                     val_in,
+                                     ind_in,
+                                     elem_count_,
+                                     bit_offset,
+                                     local_size_,
+                                     local_hist_count_,
+                                     hist_scan_deps);
+        }
+    }
+
+    sort_event_ = res_deps;
+    return res_deps;
+}
+
+template <typename Integer>
+radix_sort<Integer>::radix_sort(sycl::queue& queue, std::int64_t vector_count)
+        : queue_(queue),
+          vector_count_(0) {
+    init(queue, vector_count);
+}
+
+template <typename Integer>
+radix_sort<Integer>::~radix_sort() {
+    sort_event_.wait_and_throw();
+}
+
+template <typename Integer>
+void radix_sort<Integer>::init(sycl::queue& queue, std::int64_t vector_count) {
+    std::uint32_t uint_vector_count = de::integral_cast<std::uint32_t>(vector_count);
+    if (uint_vector_count && vector_count_ != uint_vector_count) {
```

Comment:
I don't follow the logic. If `uint_vector_count == 0` the member vars `vector_count_` and `buffer_` will be uninitialized. Shouldn't we check correctness of conversion and throw `domain_error` if conversion is impossible? You can use `dal::detail::limits` defined in `dal/detail/common.hpp`

Reply:
Added additional asserts to check for ranges, this is primitive and checks with exceptions should be done in high level func

Reply:
Hmm, I do not see assertions in `radix_sort<Integer>::radix_sort` that calls `init`.

> this is primitive and checks with exceptions should be done in high level func

Well, how the caller of this function knows about conversion from `std::int64_t` -> `std::int32_t`? The interface accepts `std::int64_t`, I don't know what happens under the hood.

```c++
void sort(..., std::int64_t);

std::int64_t vector_count= ...;
check_if_value_fits_uint32_t(vector_count);
sort(..., vector_count);
```

Nobody knows that `check_if_value_fits_uint32_t` must be called, it's too implicit contract. I'm fine with assertions for dimensions, this is easier to maintain for the caller. In this particular case, I'd suggest exceptions for integer cast. Moreover, it will simplify your RF code when you'll need to call this primitive.